File: python/222. Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        return self.helper(root)
    def helper(self, root):
        if not root:
            return 0
        r = l = root
        llevel = rlevel = 0
        while l:
            llevel += 1
            l = l.left
        while r:
            rlevel += 1
            r = r.right
        if llevel == rlevel:
            return 2 ** llevel - 1
        return self.helper(root.left) + self.helper(root.right) + 1

File: cpp/201-250/222. Count Complete Tree Nodes.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if(!root)
            return 0;
        int leftcnt = 0, rightcnt = 0;
        auto l = root;
        auto r = root;
        while(l){
            l = l->left;
            leftcnt++;
        }
        while(r){
            r = r->right;
            rightcnt++;
        }
        if(leftcnt == rightcnt)
            return pow(2, leftcnt) - 1;
        return 1 + countNodes(root->left) + countNodes(root->right);
    }
};